pkgs/_macro_tool/lib/macro_tool.dart

```diff
@@ -0,0 +1,147 @@
+        [
+          'run',
+          '--enable-experiment=macros',
+          '--enable-experiment=enhanced-parts',
+          '--packages=$packageConfigPath',
+          scriptPath
+        ],
+        workingDirectory: workspacePath,
+      );
+      stdout.write(result.stdout);
+      stderr.write(result.stderr);
+      exitCode = result.exitCode;
+    } finally {
+      if (skipCleanup) {
+        print('~~~ done running, skipping cleanup because --skip-cleanup');
+      } else {
+        print('~~~ done running, cleanup follows');
+        _removeImportAugment();
+        _removeAugmentations();
+      }
+    }
+
+    // The analyzer seems to prevent exit.
+    exit(0);
+  }
+
+  /// The path where macro-generated augmentations will be written.
+  String get _augmentationFilePath => '$scriptPath.macro_tool_output';
+
+  /// Runs macros in [scriptFile].
+  ///
+  /// Writes any augmentation to [_augmentationFilePath].
+  ///
+  /// Returns whether an augmentation file was written.
+  Future<bool> _augmentUsingAnalyzer() async {
+    final contextCollection =
+        AnalysisContextCollection(includedPaths: [workspacePath]);
+    final analysisContext = contextCollection.contexts.first;
+    injected.macroImplementation = await AnalyzerMacroImplementation.start(
+        protocol: Protocol(
+            encoding: ProtocolEncoding.binary,
+            version: ProtocolVersion.macros1),
+        packageConfig: File(packageConfigPath).uri);
+
+    final resolvedLibrary = (await analysisContext.currentSession
+        .getResolvedLibrary(scriptPath)) as ResolvedLibraryResult;
+
+    final errors = (await analysisContext.currentSession.getErrors(scriptPath))
+        as ErrorsResult;
+    if (errors.errors.isNotEmpty) {
+      print('Errors: ${errors.errors}');
+    }
+
+    final augmentationUnits =
+        resolvedLibrary.units.where((u) => u.isMacroAugmentation).toList();
+    if (augmentationUnits.isEmpty) {
+      return false;
+    }
+
+    print('Macro output (patched to use augment library): '
+        '$_augmentationFilePath');
+    File(_augmentationFilePath).writeAsStringSync(augmentationUnits
+        .first.content
+        // The analyzer produces augmentations in parts, but the CFE still
+        // wants them in augmentation libraries. Adjust the output accordingly.
+        .replaceAll('part of', 'augment library'));
+
+    return true;
+  }
+
+  /// Deletes the augmentation file created by this tool.
+  void _removeAugmentations() {
+    print('Deleting: $_augmentationFilePath');
+    File(_augmentationFilePath).deleteSync();
+  }
+
+  /// The path where the script file will be backed up before modification.
+  String get _backupScriptPath => '$scriptPath.backup';
+
+  /// Adds `import augment` of the augmentation file.
+  ///
+  /// When macros run in the analyzer or CFE this inclusion of the augmentation
+  /// output is automatic, but for `macro_tool` it has to be patched in.
+  void _addImportAugment() {
```

Comment:
Instead of overwriting the actual file, which can leave you in a bad state where you have to manually copy it back, why not just add the import to the new file and execute that?

Then, worst case we just don't do the cleanup.

Reply:
Hmm, I don't follow.

I think the file with the macro application needs the import? I don't see how to keep it in just the new file.